internal/catalog/internal/controllers/endpoints/controller.go

```diff
@@ -0,0 +1,482 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package endpoints
+
+import (
+	"context"
+	"sort"
+
+	"github.com/hashicorp/consul/internal/catalog/internal/controllers/workloadhealth"
+	"github.com/hashicorp/consul/internal/catalog/internal/types"
+	"github.com/hashicorp/consul/internal/controller"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+const (
+	endpointsMetaManagedBy = "managed-by"
+)
+
+func ServiceEndpointsController(workloadMap WorkloadMapper) controller.Controller {
+	if workloadMap == nil {
+		panic("No WorkloadMapper was provided to the ServiceEndpointsController constructor")
+	}
+
+	return controller.ForType(types.ServiceType).
+		WithWatch(types.ServiceEndpointsType, controller.MapOwner).
+		WithWatch(types.WorkloadType, workloadMap.MapWorkloadToServices).
+		WithReconciler(newServiceEndpointsReconciler(workloadMap))
+}
+
+type serviceEndpointsReconciler struct {
+	workloadMap WorkloadMapper
+
+	// We keep track of which services are being managed so that we can know
+	// when one transitions from being managed to unmanaged and delete the
+	// corresponding ServiceEndpoints accordingly.
+	managedServices map[string]struct{}
+}
+
+func newServiceEndpointsReconciler(workloadMap WorkloadMapper) *serviceEndpointsReconciler {
+	return &serviceEndpointsReconciler{
+		workloadMap:     workloadMap,
+		managedServices: make(map[string]struct{}),
+	}
+}
+
+// reconcileContext is a small struct to hold onto the varous bits of
+// data needed commonly amongst many of the functions that the reconciler
+// will call. This prevents needing to pass all these values through most
+// of the call stack.
+//
+// The notable exception is the context.Context but the docs for that
+// advise against storing those in any structure.
```

Comment:
As an aside, I think it would be acceptable under the Go guidance to have a context on this struct as long as you're only using it as a method parameter and not storing it long-term.

Another pattern I've seen before (see [this Slack thread](https://go.hashi.co/go-context-embedding) for pros/cons) is to _embed_ a context in your struct so you can pass it around, to the gRPC client for example, as if it were a regular context.

```Go
type myContext struct {
    context.Context

    someOtherField string
}

func doSomething(ctx myContext, rt controller.Runtime) {
    rsp, err := rt.Client.Read(ctx, ...)
}
```

Folks in the thread were pretty polarised on this though. So maybe best avoided!

Reply:
I ended up going a different way that didn't use this object so the whole issue was avoided.